anvil/src/shell.rs

```diff
@@ -111,21 +201,62 @@ pub fn init_shell(
-            if let ShellRequest::SetKind {
-                surface,
-                kind: ShellSurfaceKind::Toplevel,
-            } = req
-            {
-                // place the window at a random location in the [0;800]x[0;800] square
-                use rand::distributions::{Distribution, Uniform};
-                let range = Uniform::new(0, 800);
-                let mut rng = rand::thread_rng();
-                let x = range.sample(&mut rng);
-                let y = range.sample(&mut rng);
-                surface.send_configure((0, 0), wl_shell_surface::Resize::None);
-                shell_window_map
-                    .borrow_mut()
-                    .insert(SurfaceKind::Wl(surface), (x, y));
+            match req {
+                ShellRequest::SetKind {
+                    surface,
+                    kind: ShellSurfaceKind::Toplevel,
+                } => {
+                    // place the window at a random location in the [0;800]x[0;800] square
+                    use rand::distributions::{Distribution, Uniform};
+                    let range = Uniform::new(0, 800);
+                    let mut rng = rand::thread_rng();
+                    let x = range.sample(&mut rng);
+                    let y = range.sample(&mut rng);
+                    surface.send_configure((0, 0), wl_shell_surface::Resize::None);
+                    shell_window_map
+                        .borrow_mut()
+                        .insert(SurfaceKind::Wl(surface), (x, y));
+                }
+                ShellRequest::Move {
+                    surface,
+                    seat,
+                    serial,
+                } => {
+                    let seat = Seat::from_resource(&seat).unwrap();
+                    // TODO: touch move.
+                    let pointer = seat.get_pointer().unwrap();
+
+                    // Check that this surface has a click grab.
+                    if !pointer.has_grab(serial) {
+                        return;
+                    }
+
+                    let start_data = pointer.grab_start_data().unwrap();
+
+                    // If the focus was for a different surface, ignore the request.
+                    if start_data.focus.is_none()
+                        || !compositor_token.contains_in_tree(
+                            surface.get_surface().unwrap(),
+                            &start_data.focus.as_ref().unwrap().0,
+                        )
+                    {
+                        return;
+                    }
+
+                    let toplevel = SurfaceKind::Wl(surface);
+                    let initial_window_location = shell_window_map.borrow().location(&toplevel).unwrap();
+
+                    let grab = MoveSurfaceGrab {
+                        start_data,
+                        window_map: shell_window_map.clone(),
+                        toplevel,
+                        // TODO: use location at the time of the event with the given serial.
```

Comment:
Tbh I don't think this is worth doing in anvil, or at all.

We don't want to store a large history of what occurred just because some client decided to wait a long time between the start of its implicit grab and the moment it sent a `move` request...

Reply:
Maybe we can re-purpose the surface location part of the focus to be the toplevel location and then just use that?

Reply:
The "location of the toplevel surface to which this surface belongs at the start of the drag" is imo way too specific to be something stored at this level. Would it be of any use to anything other than the move grab?

Reply:
Yeah, I'm not sure about this. I suppose that focus location can be just removed for now, and, should it become really useful, a better mechanism can be made.

Reply:
Ah, actually it's used in `motion` to get the surface-relative coordinates. My bad, didn't notice.